### ISSUE TITLE
large allocation in index_consumer

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2710]
*Installation details*
Scylla version (or git commit hash): 2.1

512kB.

```
seastar::current_backtrace()
/home/avi/urchin/seastar/core/reactor.cc:3943
seastar::memory::cpu_pages::warn_large_allocation(unsigned long)
/home/avi/urchin/seastar/core/memory.cc:627
seastar::memory::cpu_pages::allocate_large(unsigned int)
crtstuff.c:?
seastar::memory::allocate(unsigned long)
crtstuff.c:?
operator new(unsigned long)
??:?
__gnu_cxx::new_allocator<sstables::index_entry>::allocate(unsigned long, void const*)
/usr/include/c++/7/ext/new_allocator.h:111
std::allocator_traits<std::allocator<sstables::index_entry> >::allocate(std::allocator<sstables::index_entry>&, unsigned long)
/usr/include/c++/7/bits/alloc_traits.h:436
std::_Vector_base<sstables::index_entry, std::allocator<sstables::index_entry> >::_M_allocate(unsigned long)
/usr/include/c++/7/bits/stl_vector.h:172
void std::vector<sstables::index_entry, std::allocator<sstables::index_entry> >::_M_realloc_insert<sstables::index_entry>(__gnu_cxx::__normal_iterator<sstables::index_entry*, std::vector<sstables::index_entry, std::allocator<sstables::index_entry> > >, sstables::index_entry&&)
/usr/include/c++/7/bits/vector.tcc:406
data_consumer::continuous_data_consumer<sstables::index_consume_entry_context<sstables::index_consumer> >::process(seastar::temporary_buffer<char>&)
crtstuff.c:?
seastar::future<> seastar::input_stream<char>::consume<sstables::index_consume_entry_context<sstables::index_consumer> >(sstables::index_consume_entry_context<sstables::index_consumer>&)::{lambda()#1}::operator()() const
crtstuff.c:?
seastar::future<> seastar::repeat<seastar::future<> seastar::input_stream<char>::consume<sstables::index_consume_entry_context<sstables::index_consumer> >(sstables::index_consume_entry_context<sstables::index_consumer>&)::{lambda()#1}>(sstables::index_consume_entry_context<sstables::index_consumer>&&)
crtstuff.c:?
seastar::continuation<seastar::future<> seastar::future<seastar::bool_class<seastar::stop_iteration_tag> >::then<seastar::future<> seastar::repeat<seastar::future<> seastar::input_stream<char>::consume<sstables::index_consume_entry_context<sstables::index_consumer> >(sstables::index_consume_entry_context<sstables::index_consumer>&)::{lambda()#1}>(sstables::index_consume_entry_context<sstables::index_consumer>&&)::{lambda(seastar::bool_class<seastar::stop_iteration_tag>)#1}, seastar::future<> >(seastar::future<> seastar::input_stream<char>::consume<sstables::index_consume_entry_context<sstables::index_consumer> >(sstables::index_consume_entry_context<sstables::index_consumer>&)::{lambda()#1})::{lambda(seastar::future<> seastar::input_stream<char>::consume<sstables::index_consume_entry_context<sstables::index_consumer> >(sstables::index_consume_entry_context<sstables::index_consumer>&)::{lambda()#1})#1}, seastar::bool_class<seastar::stop_iteration_tag> >::run()
crtstuff.c:?
seastar::reactor::run_tasks(seastar::circular_buffer<std::unique_ptr<seastar::task, std::default_delete<seastar::task> >, std::allocator<std::unique_ptr<seastar::task, std::default_delete<seastar::task> > > >&)
crtstuff.c:?
seastar::reactor::run()
crtstuff.c:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&)
/home/avi/urchin/seastar/core/app-template.cc:142
main
/home/avi/urchin/main.cc:696
??
??:0
_start
??:?
```

**Comment by slivne** [comment]
@avikivity - won't the work from @argenet fix this ?

**Comment by argenet** [comment]
It seems this particular one happens on parsing/inserting index_entries which is going to remain the same with the current approach to #2981. We have tried an approach where both index_entries and promoted index blocks were parsed incrementally but found it to hit badly the cases when we cannot properly estimate the upper bound for a partition range read.

**Comment by avikivity** [comment]
Fixed by 3b2890e1db770b0d23b723df0843073f0cab1a01.
